pkl-core/src/main/java/org/pkl/core/stdlib/ini/RendererNodes.java

```diff
@@ -0,0 +1,409 @@
+
+    @Override
+    protected void visitEntryKey(Object key, boolean isFirst) {
+      storedIsFirst = isFirst;
+      if (VmUtils.isRenderDirective(key)) {
+        visitRenderDirective((VmTyped) key);
+        writeSeparator();
+        return;
+      }
+
+      if (key instanceof String) {
+        storedEntryKey = (String) key;
+        return;
+      }
+
+      cannotRenderNonStringKey(key);
+    }
+
+    @Override
+    protected void visitEntryValue(Object value) {
+      if (!((value instanceof VmTyped)
+          || (value instanceof VmDynamic)
+          || (value instanceof VmMapping)
+          || (value instanceof VmMap))) {
+        if (!storedIsFirst) {
+          writeLineBreak();
+        }
+        writeKeyOrValue(storedEntryKey);
+        writeSeparator();
+        visit(value);
+      }
+    }
+
+    @Override
+    protected void visitProperty(Identifier name, Object value, boolean isFirst) {
+
+      if (!((value instanceof VmTyped)
+          || (value instanceof VmDynamic)
+          || (value instanceof VmMapping)
+          || (value instanceof VmMap))) {
+        if (!isFirst) {
+          writeLineBreak();
+        }
+        writeKeyOrValue(name.toString());
+        writeSeparator();
+        visit(value);
+      }
+    }
+
+    /*
+    These end functions group `VmTyped`, `VmMapping`, `VmDynamic`, `VmMap` at the end of the map and then writes them to builder
+    This is to comply with ini style standards. Please look at tests for example's
+     */
+    @Override
+    protected void endDynamic(VmDynamic value, boolean isEmpty) {
+      value.forceAndIterateMemberValues(
+          ((memberKey, member, memberValue) -> {
+            if ((memberValue instanceof VmTyped)
+                || (memberValue instanceof VmDynamic)
+                || (memberValue instanceof VmMapping)
+                || (memberValue instanceof VmMap)) {
+
+              if (memberKey instanceof Identifier) {
+                currPath.push(memberKey);
+              } else {
+                currPath.push(converter.convert(memberKey, List.of()));
+              }
+              visit(memberValue);
+              currPath.pop();
+            }
+            return true;
+          }));
+    }
+
+    @Override
+    protected void endTyped(VmTyped value, boolean isEmpty) {
+      value.forceAndIterateMemberValues(
+          ((memberKey, member, memberValue) -> {
+            if ((memberValue instanceof VmTyped)
+                || (memberValue instanceof VmDynamic)
+                || (memberValue instanceof VmMapping)
+                || (memberValue instanceof VmMap)) {
+
+              if (memberKey instanceof Identifier) {
+                currPath.push(memberKey);
+              } else {
+                currPath.push(converter.convert(memberKey, List.of()));
+              }
+              visit(memberValue);
+              currPath.pop();
+            }
+            return true;
+          }));
```

Comment:
Should have caught this earlier; this is where your bug is. You already print the contents of an object by visiting each property _before_ this method gets called (that provides the expected output). This is the bit that produces the unexpected extra output. The `endDynamic` / `endTyped` / `endMap` / etc methods are there for other kinds of admin, such as decreasing indent (in JSON, YAML, Pcf) and printing a closing `}` (in JSON, Pcf). You shouldn't revisit the members.

Reply:
Ah, Thank you. I understand where you are coming from, at the time i though it would be a good idea to use them for complex Pkl files. See the issue comees when you have a pkl file like this for example:
```pkl
mapping = new Mapping{
  ["one"] = "value1"
  ["two"] = new Person {name = "Pigeon" age = 19 address {
    street = "fun Street"
  }}
  ["three"] = "value3"
  ["four"] = 1.2
  ["five"] = new Mapping {
    ["six"] = 123
  }
  ["seven"] = true
}
```
The expected output would have the correct sections such as `mapping.one` or `mapping.two.address` and then the values under said sections. For example:

```pkl
[mapping]
one = value1
three = value3
four = 1.2
seven = true

[mapping.two]
name = Pigeon
age = 19

[mapping.two.address]
street = fun Street

[mapping.five]
six = 123
```

So the use of these "end functions" is to make sure that the values are correctly under each section. If i didnt use the end functions the rendered ini file would look like this: 
```pkl
[mapping]
one = value1

[mapping.two]
name = Pigeon
age = 19

[mapping.two.address]
street = fun Street
three = value3
four = 1.2

[mapping.five]
six = 123
seven = true
```
And as you can see some values such as `seven` and `four` are under the incorrect section.

Im not sure if you or anyone else could think of another way of overcoming this problem if so it would be great to get this solved!

Reply:
@holzensp have you had time to look at my issue?

Reply:
Again, sluggish; sorry. I get it now; you're having to compensate for `.ini` not having hierarchies, so things that would be nested, now come _after_ the object is entirely rendered. You should still not _revisit the object's members_, though. I'd keep a stack of `StringWriter`s in your renderer. Every `startFoo` you push a new one with the qualified name of the property already in it. Every `endFoo` you then pop it and write the top-writers contents.

Reply:
Thank you so much! Ill have a look soon